utils_nlp/models/transformers/common.py

```diff
@@ -34,127 +31,192 @@
-    per_gpu_train_batch_size=8,
-    n_gpu=1,
-    weight_decay=0.0,
-    learning_rate=5e-5,
-    adam_epsilon=1e-8,
-    warmup_steps=0,
-    fp16=False,
-    fp16_opt_level="O1",
-    local_rank=-1,
-    seed=0,
-):
-
-    set_seed(seed, n_gpu > 0)
-
-    train_batch_size = per_gpu_train_batch_size * max(1, n_gpu)
-    train_sampler = (
-        RandomSampler(train_dataset) if local_rank == -1 else DistributedSampler(train_dataset)
-    )
-    train_dataloader = DataLoader(train_dataset, sampler=train_sampler, batch_size=train_batch_size)
-
-    if max_steps > 0:
-        t_total = max_steps
-        num_train_epochs = max_steps // (len(train_dataloader) // gradient_accumulation_steps) + 1
+def get_device(device, num_gpus, local_rank):
+    if local_rank == -1:
+        device = torch.device("cuda" if torch.cuda.is_available() and device == "cuda" else "cpu")
+        num_gpus = (
+            min(num_gpus, torch.cuda.device_count()) if num_gpus else torch.cuda.device_count()
+        )
     else:
-        t_total = len(train_dataloader) // gradient_accumulation_steps * num_train_epochs
-
-    no_decay = ["bias", "LayerNorm.weight"]
-    optimizer_grouped_parameters = [
-        {
-            "params": [
-                p for n, p in model.named_parameters() if not any(nd in n for nd in no_decay)
-            ],
-            "weight_decay": weight_decay,
-        },
-        {
-            "params": [p for n, p in model.named_parameters() if any(nd in n for nd in no_decay)],
-            "weight_decay": 0.0,
-        },
-    ]
-    optimizer = AdamW(optimizer_grouped_parameters, lr=learning_rate, eps=adam_epsilon)
-    scheduler = WarmupLinearSchedule(optimizer, warmup_steps=warmup_steps, t_total=t_total)
-
-    if fp16:
-        try:
-            from apex import amp
-        except ImportError:
-            raise ImportError(
-                "Please install apex from https://www.github.com/nvidia/apex to use fp16 training."
-            )
-        model, optimizer = amp.initialize(model, optimizer, opt_level=fp16_opt_level)
+        torch.cuda.set_device(local_rank)
+        device = torch.device("cuda", local_rank)
+        torch.distributed.init_process_group(backend="nccl")
+        num_gpus = 1
 
-    # multi-gpu training (should be after apex fp16 initialization)
-    if n_gpu > 1:
-        model = torch.nn.DataParallel(model)
+    return device, num_gpus
 
-    # Distributed training (should be after apex fp16 initialization)
-    if local_rank != -1:
-        model = torch.nn.parallel.DistributedDataParallel(
-            model, device_ids=[local_rank], output_device=local_rank, find_unused_parameters=True
+
+class Transformer:
+    def __init__(self, model_class, model_name="bert-base-cased", cache_dir=".", seed=0, **kwargs):
+        self.model_name = model_name
+        self.model = model_class[model_name].from_pretrained(
+            model_name, cache_dir=cache_dir, num_labels=kwargs["num_labels"]
```

Comment:
Why having "num_labels" in kwargs? If people don't provide it when calling __init__, a key error will be raise, Then what's the difference compared to making num_labels an explicit, required argument?

Reply:
I think I'll just remove all kwargs. I was initially thinking we can link to the transformers documentation and let the users decide what to pass.